plugins/crypto/src/main/java/org/opensearch/cryptoplugin/CryptoBasePlugin.java

```diff
@@ -0,0 +1,168 @@
+        final NodeEnvironment nodeEnvironment,
+        final NamedWriteableRegistry namedWriteableRegistry,
+        final IndexNameExpressionResolver expressionResolver,
+        final Supplier<RepositoriesService> repositoriesServiceSupplier
+    ) {
+        cacheRefresher = new CacheRefresher(CRYPTO_ASYNC_REFRESH_THREADPOOL, threadPool);
+        return Collections.emptyList();
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public CryptoManager.Factory createClientFactory(String keyProviderType) {
+        return (cryptoSettings, keyProviderName) -> {
+            if (!keyProviderExtensions.containsKey(keyProviderType)) {
+                throw new IllegalArgumentException(
+                    "Key provider extension of type " + keyProviderType + " not found in the installed crypto extension plugins."
+                );
+            }
+
+            synchronized (keyProviderCryptoManagers) {
+                Map<String, CryptoManagerImpl> existingCryptoStores = keyProviderCryptoManagers.get(keyProviderType);
+                if (existingCryptoStores != null && existingCryptoStores.containsKey(keyProviderName)) {
+                    existingCryptoStores.get(keyProviderName).incRef();
+                    return existingCryptoStores.get(keyProviderName);
+                }
+
+                MasterKeyProvider masterKeyProvider = keyProviderExtensions.get(keyProviderType).createKeyProvider(cryptoSettings);
+                Map<String, String> encryptionContext = new HashMap<>();
+                CryptoAlgorithm cryptoAlgorithm = getCryptoAlgorithm();
+                MasterKeyCache masterKeyCache = new MasterKeyCache(
+                    cryptoAlgorithm,
+                    cacheTTL,
+                    masterKeyProvider,
+                    keyProviderName,
+                    encryptionContext
+                );
+                cacheRefresher.register(masterKeyCache);
+                CryptoManagerImpl cryptoManagerImpl = new CryptoManagerImpl(keyProviderName, keyProviderType, () -> {
+                    cacheRefresher.deregister(masterKeyCache);
+                    keyProviderCryptoManagers.get(keyProviderType).remove(keyProviderName);
+                    if (keyProviderCryptoManagers.get(keyProviderType).isEmpty() == true) {
+                        keyProviderCryptoManagers.remove(keyProviderType);
+                    }
+                }, masterKeyCache, encryptionContext, cryptoAlgorithm);
+                cryptoManagerImpl.incRef();
+                keyProviderCryptoManagers.putIfAbsent(keyProviderType, new HashMap<>());
+                keyProviderCryptoManagers.get(keyProviderType).put(keyProviderName, cryptoManagerImpl);
+            }
+            return keyProviderCryptoManagers.get(keyProviderType).get(keyProviderName);
+        };
+    }
+
+    CryptoAlgorithm getCryptoAlgorithm() {
+        return CryptoAlgorithm.ALG_AES_256_GCM_HKDF_SHA512_COMMIT_KEY;
```

Comment:
Can we make this configurable with a default?

Reply:
Made this a node level static setting.